File: WindowsApplication3c/MainForm.vb
Imports System.Net
Imports System.Text.RegularExpressions
Imports System.Text
Imports System.IO
Imports System.ComponentModel

Public Class MainForm
    Dim _webRequest As HttpWebRequest

    Private Sub getHtmlButton_Click(sender As Object, e As EventArgs) Handles getHtmlButton.Click
        listBox.Items.Clear()
        _webRequest = DirectCast(WebRequest.Create("http://www.gutenberg.org/files/1497/1497.txt"), HttpWebRequest)
        BackgroundWorker.RunWorkerAsync(_webRequest)
        getHtmlButton.Enabled = False
        progressBar.Value = 0
        progressBar.Visible = True
        cancelButton.Visible = True
    End Sub

    Private Sub cancelButton_Click(sender As Object, e As EventArgs) Handles cancelButton.Click
        Dim request As HttpWebRequest = _webRequest
        If (Not request Is Nothing) Then
            BackgroundWorker.CancelAsync()
            request.Abort()
        End If
    End Sub

    Private Shared Function GetScriptBodies(ByVal html As String) As IEnumerable(Of String)
        Const pattern As String = "<script.*?>\s*(?'scriptBody'.+?)\s*</script>"
        Dim matches As MatchCollection = Regex.Matches(html, pattern)
        Dim result As List(Of String) = New List(Of String)(matches.Count)
        For Each match As Match In matches
            Dim scriptBody As String = match.Groups("scriptBody").Value
            result.Add(scriptBody)
        Next
        Return result
    End Function

    Private Sub BackgroundWorker_DoWork(sender As Object, e As System.ComponentModel.DoWorkEventArgs) _
        Handles BackgroundWorker.DoWork
        Dim progress As Integer = 1
        Dim backgroundWokrer As BackgroundWorker = DirectCast(sender, BackgroundWorker)
        backgroundWokrer.ReportProgress(progress)
        Dim request As HttpWebRequest = DirectCast(e.Argument, HttpWebRequest)
        Try
            Dim response As WebResponse = request.GetResponse()
            Dim stream As Stream = response.GetResponseStream()
            Dim numArray(1024000) As Byte
            Dim asyncResult As IAsyncResult = StreamHelper.
                    BeginReadStreamToEnd(stream, numArray, 0,
                                         CInt(numArray.Length),
                                         Sub(ar As IAsyncResult)
                                         Dim bytesRead As Integer = StreamHelper.EndReadStreamToEnd(ar)
                                         Array.Resize(Of Byte)(numArray, bytesRead)
                                     End Sub, stream)

            While False = asyncResult.AsyncWaitHandle.WaitOne(200)
                BackgroundWorker.ReportProgress(progress)
                progress += (100 - progress) / 2
                If (Not BackgroundWorker.CancellationPending) Then Continue While
                e.Cancel = True
                Return
            End While
            backgroundWokrer.ReportProgress(100)
            e.Result = Encoding.ASCII.GetString(numArray, 0, CInt(numArray.Length))
        Catch ex As WebException
            If Not ex.Status = WebExceptionStatus.RequestCanceled Then Throw
            e.Cancel = True
        End Try
    End Sub

    Private Sub BackgroundWorker_ProgressChanged(sender As Object, e As System.ComponentModel.ProgressChangedEventArgs) Handles BackgroundWorker.ProgressChanged
        progressBar.Value = e.ProgressPercentage
    End Sub

    Private Sub BackgroundWorker_RunWorkerCompleted(sender As Object, e As System.ComponentModel.RunWorkerCompletedEventArgs) Handles BackgroundWorker.RunWorkerCompleted
        getHtmlButton.Enabled = True
        cancelButton.Visible = False
        _webRequest = Nothing
        If (e.Cancelled) Then
            MessageBox.Show("Cancelled")
            progressBar.Visible = False
        Else
            For Each x As String In GetScriptBodies(e.Result)
                listBox.Items.Add(x)
            Next
        End If

    End Sub
End Class

File: WindowsFormsApplication3c/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Windows.Forms;

namespace WindowsFormsApplication3c
{
	public partial class MainForm : Form
	{
		public MainForm()
		{
			InitializeComponent();
		}

		private HttpWebRequest webRequest;

		private void getHtmlButton_Click(object sender, EventArgs e)
		{
			listBox.Items.Clear();

			backgroundWorker.RunWorkerAsync();

			getHtmlButton.Enabled = false;
			progressBar.Value = 0;
			progressBar.Visible = true;
			cancelButton.Visible = true;
		}

		private void backgroundWorker_DoWork(object sender, DoWorkEventArgs e)
		{
			int progress = 1;
			var backgroundWorker = (BackgroundWorker) sender;
			backgroundWorker.ReportProgress(progress);
			webRequest =
				(HttpWebRequest)WebRequest.Create("http://www.gutenberg.org/files/1497/1497.txt");
			try
			{
				var response = webRequest.GetResponse();
				var stream = response.GetResponseStream();
				byte[] inputBuffer = new byte[1024000];
				var asyncResult = StreamHelper.
					BeginReadStreamToEnd(stream, inputBuffer, 0, inputBuffer.Length,
										 ar =>
											 {
												 int bytesRead = StreamHelper.EndReadStreamToEnd(ar);
												 Array.Resize(ref inputBuffer, bytesRead);
												 ((IDisposable) response).Dispose();
											 }, stream);

				while (false == asyncResult.AsyncWaitHandle.WaitOne(200))
				{
					backgroundWorker.ReportProgress(progress);
					progress += (100 - progress)/2;
					if (!backgroundWorker.CancellationPending) continue;
					e.Cancel = true;
					return;
				}
				backgroundWorker.ReportProgress(100);
				e.Result = Encoding.ASCII.GetString(inputBuffer, 0, inputBuffer.Length);
			}
			catch (WebException ex)
			{
				if (ex.Status != WebExceptionStatus.RequestCanceled) throw;
				e.Cancel = true;
			}
		}

		private void backgroundWorker_ProgressChanged(object sender, ProgressChangedEventArgs e)
		{
			progressBar.Value = e.ProgressPercentage;
		}

		private void backgroundWorker_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
		{
			getHtmlButton.Enabled = true;
			cancelButton.Visible = false;
			webRequest = null;
			if (e.Cancelled)
			{
				MessageBox.Show("Cancelled");
				progressBar.Visible = false;
			}
			else
			{
				foreach (string x in GetScriptBodies((string) e.Result))
					listBox.Items.Add(x);
			}
		}

		private void cancelButton_Click(object sender, EventArgs e)
		{
			var request = webRequest;
			if (request == null) return;
			backgroundWorker.CancelAsync();
			request.Abort();
		}

		private static IEnumerable<string> GetScriptBodies(string html)
		{
			const string pattern = @"<script.*?>\s*(?'scriptBody'.+?)\s*</script>";
			if (html != null)
			{
				MatchCollection matches = Regex.Matches(html, pattern);
				if (matches.Count == 0) yield break;
				foreach (Match match in matches)
				{
					string scriptBody = match.Groups["scriptBody"].Value;
					yield return scriptBody;
				}
			}
			yield break;
		}
	}
}
